File: 61-rotate-list/rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
                if (head == null || head.next == null || k == 0) {
            return head;
        }
        
        ListNode current = head;
        int length = 1;
        while (current.next != null) {
            current = current.next;
            length++;
        }
        
        k = k % length;
        if (k == 0) {
            return head;
        }
        
        current.next = head; // Make it a circular linked list
        int stepsToNewHead = length - k;
        ListNode newTail = head;
        for (int i = 1; i < stepsToNewHead; i++) {
            newTail = newTail.next;
        }
        
        ListNode newHead = newTail.next;
        newTail.next = null;
        return newHead;
        
    }
}

File: 61-rotate-list/rotate-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
int len(struct ListNode* head){
    int count =0;
    while(head!=NULL){
        count++;
        head = head->next;
    }
    return count;
}
struct ListNode* removeTail(struct ListNode* head, struct ListNode** dNode) {
    if (head == NULL) {
        *dNode = NULL;
        return NULL;
    }

    if (head->next == NULL) { // only 1 node
        *dNode = head;
        return NULL;  // list becomes empty
    }

    struct ListNode* temp = head;

    while (temp->next->next != NULL) {
        temp = temp->next;
    }

    *dNode = temp->next;     // last node
    temp->next = NULL;       // detach tail

    return head;             // updated list
}

struct ListNode* addFirst(struct ListNode* head,struct ListNode* node){
    node->next = head;
    return node;

}
struct ListNode* rotateRight(struct ListNode* head, int k) {
    int n = len(head);
    if (n == 0 || k == 0) return head;

    k %= n;
    if (k == 0) return head;

    struct ListNode* dNode = NULL;

    while (k--) {
        head = removeTail(head, &dNode);  // get last node in dNode
        dNode->next = head;               // insert at head
        head = dNode;                     // update head
    }

    return head;
}
